src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/Pileup.java

```diff
@@ -0,0 +1,176 @@
+public class Pileup extends LocusWalker {
+
+    private static final String verboseDelimiter = "@"; // it's ugly to use "@" but it's literally the only usable character not allowed in read names
+
+    @Argument(fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME, shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME, doc = "An output file created by the walker. Will overwrite contents if file exists")
+    public PrintStream out;
+
+    /**
+     * In addition to the standard pileup output, adds 'verbose' output too. The verbose output contains the number of
+     * spanning deletions, and for each read in the pileup it has the read name, offset in the base string, read length,
+     * and read mapping quality.  These per read items are delimited with an '@' character.
+     */
+    @Argument(fullName = "showVerbose", shortName = "verbose", doc = "Add an extra verbose section to the pileup output", optional = true)
+    public boolean SHOW_VERBOSE = false;
+
+    /**
+     * This enables annotating the pileup to show overlaps with metadata from a ROD file. For example, if you provide a
+     * VCF and there is a SNP at a given location covered by the pileup, the pileup output at that position will be
+     * annotated with the corresponding source ROD identifier.
+     */
+    @Argument(fullName = "metadata", shortName = "metadata", doc = "ROD file containing metadata", optional = true)
+    public List<FeatureInput<Feature>> rods = new ArrayList<>();
+
+    /**
+     * Adds the length of the insert each base comes from to the output pileup. Here, "insert" refers to the DNA insert
+     * produced during library generation before sequencing.
+     */
+    @Hidden
+    @Argument(fullName = "outputInsertLength", shortName = "outputInsertLength", doc = "Output insert length", optional = true)
+    public boolean outputInsertLength = false;
+
+    private int result = 0;
+
+    @Override
+    public void apply(AlignmentContext context, ReferenceContext ref, FeatureContext tracker) {
+        final String rods = getReferenceOrderedData(tracker);
+        final ReadPileup basePileup = context.getBasePileup();
+        final StringBuilder s = new StringBuilder();
+        s.append(
+            String.format("%s %s", basePileup.getPileupString((hasReference()) ? (char) ref.getBase() : 'N'), rods));
+        if (outputInsertLength)
+            s.append(" ").append(insertLengthOutput(basePileup));
+        if (SHOW_VERBOSE)
+            s.append(" ").append(createVerboseOutput(basePileup));
+        s.append("\n");
+        out.print(s.toString());
+        result++;
+    }
+
+    /**
+     * Get a string representation the reference-ordered data.
+     *
+     * @param tracker Container for the reference-ordered data.
+     *
+     * @return String representation of the reference-ordered data.
+     */
+    private String getReferenceOrderedData(final FeatureContext tracker) {
+        ArrayList<String> rodStrings = new ArrayList<>();
```

Comment:
declare temp using the interface `final List<String>` - the only legitimate use for `ArrayList` is in `new ArrayList<>`